plugins/kubernetes-backend/src/service/KubernetesProxy.test.ts

```diff
@@ -0,0 +1,342 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { getVoidLogger } from '@backstage/backend-common';
+import {
+  ClusterDetails,
+  KubernetesClustersSupplier,
+  KubernetesProxyServices,
+} from '../types/types';
+import { KubernetesProxy } from './KubernetesProxy';
+
+import { Request } from 'express';
+
+import 'buffer';
+
+jest.mock('node-fetch');
+const { Response } = jest.requireActual('node-fetch');
+
+import fetch from 'node-fetch';
+
+describe('KubernetesProxy', () => {
+  let _clientMock: any;
+  let sut: KubernetesProxy;
+
+  const buildEncodedRequest = (
+    clustersHeader: any,
+    query: string,
+    body?: any,
+  ): Request => {
+    const encodedQuery = encodeURIComponent(query);
+    const encodedClusters = Buffer.from(
+      JSON.stringify(clustersHeader),
+    ).toString('base64');
+
+    const req = {
+      params: {
+        encodedQuery,
+      },
+      header: (key: string) => {
+        let value: string = '';
+        switch (key) {
+          case 'Content-Type': {
+            value = 'application/json';
+            break;
+          }
+          case 'X-Kubernetes-Clusters': {
+            value = encodedClusters;
+            break;
+          }
+          default: {
+            break;
+          }
+        }
+        return value;
+      },
+    } as unknown as Request;
+
+    if (body) {
+      req.body = body;
+    }
+
+    return req;
+  };
+
+  const buildProxyServicesWithClusters = (
+    clusters: ClusterDetails[],
+  ): KubernetesProxyServices => {
+    const kcs: KubernetesClustersSupplier = {
+      getClusters: async () => {
+        return clusters;
+      },
+    };
+
+    return {
+      kcs,
+    };
+  };
+
+  beforeEach(() => {
+    jest.resetAllMocks();
+    _clientMock = {
+      handleProxyRequest: jest.fn(),
+    };
+
+    sut = new KubernetesProxy(getVoidLogger());
+  });
+
+  it('should return a 404 if no clusters are found', async () => {
+    const services = buildProxyServicesWithClusters([]);
+    const req = buildEncodedRequest({}, 'api');
+
+    const result = await sut.handleProxyRequest(services, req);
+
+    expect(result.code).toEqual(404);
+    expect(fetch).not.toHaveBeenCalled();
+  });
+
+  it('should match the response code of the Kubernetes response (single cluster)', async () => {
+    const services = buildProxyServicesWithClusters([
+      {
+        name: 'cluster1',
+        url: 'http://localhost:9999',
+        serviceAccountToken: 'token',
+        authProvider: 'serviceAccount',
+        skipTLSVerify: true,
+      },
+    ]);
+    const req = buildEncodedRequest({ cluster1: 'token' }, 'api');
+
+    const apiResponse = {
+      kind: 'APIVersions',
+      versions: ['v1'],
+      serverAddressByClientCIDRs: [
+        {
+          clientCIDR: '0.0.0.0/0',
+          serverAddress: '192.168.0.1:3333',
+        },
+      ],
+    };
+
+    // @ts-ignore-next-line
```

Comment:
yeah in general try very hard to avoid `@ts-ignore` - with MSW as mentioned above and actual request/response cycles being exercised, you will both get rid of this exemption and get a more solid test in the end

Reply:
Done